minidump-processor/src/op_analysis.rs

```diff
@@ -0,0 +1,733 @@
+/// this function
+///
+pub fn pretty_print_thread_instruction(
+    context: &MinidumpContext,
+    memory: &mut impl SparseMemoryStream,
+) -> Result<String, OpAnalysisError> {
+    let instruction_bytes = get_thread_instruction_bytes(context, memory)?;
+    pretty_print_instruction_bytes(context, &instruction_bytes)
+}
+
+/// Analysis tools for the Amd64 architecture
+#[cfg(feature = "disasm_amd64")]
+mod amd64 {
+    use super::*;
+
+    /// Amd64-specific `get_instruction_memory_access`. See docs for general function
+    pub fn get_instruction_memory_access(
+        context: &MinidumpContext,
+        bytes: &[u8],
+    ) -> Result<Vec<MemoryAccess>, OpAnalysisError> {
+        use yaxpeax_x86::amd64::{DecodeError, InstDecoder, Operand, RegSpec};
+
+        let calculate_address = |base_reg: Option<RegSpec>,
+                                 index_reg: Option<RegSpec>,
+                                 scale: Option<u8>,
+                                 disp: Option<i32>|
+         -> Result<u64, OpAnalysisError> {
+            let get_reg = |reg: RegSpec| -> Result<u64, OpAnalysisError> {
+                context
+                    .get_register(reg.name())
+                    .ok_or(OpAnalysisError::RegisterInvalid)
+            };
+
+            let base = match base_reg {
+                Some(reg) => get_reg(reg)?,
+                None => 0,
+            };
+
+            let scaled_index = match index_reg {
+                Some(reg) => {
+                    let index = get_reg(reg)?;
+                    let scale = scale.unwrap_or(1);
+                    index.wrapping_mul(scale.into())
+                }
+                None => 0,
+            };
+
+            let disp = disp.unwrap_or(0) as i64 as u64;
+
+            Ok(base.wrapping_add(scaled_index).wrapping_add(disp))
+        };
+
+        let decoder = InstDecoder::default();
+        let decoded_instruction = decoder.decode_slice(bytes).map_err(|error| match error {
+            DecodeError::ExhaustedInput => OpAnalysisError::InstructionTruncated,
+            e => OpAnalysisError::DecodeFailed(e.into()),
+        })?;
+
+        let mut memory_accesses = Vec::new();
+
+        // Shortcut -- If the instruction doesn't access memory, just return an empty list
+        let mem_size = match decoded_instruction.mem_size() {
+            Some(access) => access.bytes_size(),
+            None => return Ok(memory_accesses),
+        };
+
+        for idx in 0..decoded_instruction.operand_count() {
+            let operand = decoded_instruction.operand(idx);
+
+            let maybe_address = match operand {
+                Operand::DisplacementU32(disp) => Some(disp.into()),
+                Operand::DisplacementU64(disp) => Some(disp),
+                Operand::RegDeref(base) => Some(calculate_address(Some(base), None, None, None)?),
+                Operand::RegDisp(base, disp) => {
+                    Some(calculate_address(Some(base), None, None, Some(disp))?)
```

Comment:
hmm there's a potential argument here that instead of `?` this should be `.ok()`, so that you can recover *some* memory accesses even if others fail. I think that could happen if you're looking at a frame that isn't the context (top) frame, as this would result in only some of the registers being valid.

That said the caller would need to then be wary of incorrect "negative reasoning" about the absence of memory accesses (because it could just be that some of them were silently dropped for being uncomputable).